man/man1/ngsconvert.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\"
.\" Copyright (c) 1985 Wayne A. Christopher, U. C. Berkeley CAD Group
.\"
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH SCONVERT 1 "2014-10-28"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.UC 4
.SH NAME
sconvert \- convert spice formats
.SH SYNOPSIS
.B sconvert fromtype fromfile totype tofile
.br
.B sconvert fromtype totype
.br
.B sconvert
.br
.SH DESCRIPTION
.B Sconvert
translates spice output files among three formats: the old
binary format, a new binary format, and a new ascii format.
The formats are specified by the
.B fromtype
and
.B totype
arguments: `o' for the old format, `b' for the new binary format,
and `a' for the new ascii format.
.B Fromtype
specifies the format to be read, and
.B totype
specifies the format to be written.
If
.B fromfile
and
.B tofile
are given, then they are used as the input and output, otherwise
standard input and output are used.
(Note that this second option is only available on \s-2UNIX\s+2 systems
\- on VMS and other systems you must supply the filenames.)
If no arguments are given, the parameters are prompted for.
.PP
Binary format is the preferred format for general use, as it is
the most economical in terms of space and speed of access, and ascii is
provided to make it easy to modify data files and transfer them
between machines with different floating-point formats.
The old format is provided only
for backward compatibility.
The three formats are as follows:
.br
.nf

.B Old:

	    What		   Size in Bytes

	title				80
	date				8
	time				8
	numoutputs			2
	the integer 4			2
	variable names --
		char[numoutputs][8]	numoutputs * 8
	types of output			numoutputs * 2
	node index			numoutputs * 2
	plot title			numoutputs * 24
	the actual data			numpoints * numoutputs * 8

.B Ascii:

	Title: \fITitle Card String\fR
	Date: \fIDate\fR
	[ Plotname: \fIPlot Name\fR
	  Flags: \fIcomplex\fR or \fIreal\fR
	  No. Variables: \fInumoutputs\fR
	  No. Points: \fInumpoints\fR
	  Command: \fInutmeg command\fR
	  Variables:	0 \fIvarname1\fR \fItypename1\fR
			1 \fIvarname2\fR \fItypename2\fR
			etc...
	  Values:
	    0	n	n	n	n	...
	    1	n	n	n	n	...
	    And so forth...
	] repeated one or more times
.fi
.PP
If one of the flags is \fIcomplex\fR, the points look like r,i where r and i
are floating point (in %e format).
Otherwise they are in %e format.
Only one of \fIreal\fR and \fIcomplex\fR should appear.
.PP
The lines are guaranteed to be less than 80 columns wide (unless the
plot title or variable names are very long), so this format is safe
to mail between systems like CMS.
.PP
Any number of \fBCommand:\fR lines may appear between the \fBNo. Points:\fR
and the \fBVariables:\fR lines, and whenever the plot is loaded into
\fBnutmeg\fR they will be executed.
.nf

.B Binary:

	\fITitle Card\fR		  (a NULL terminated string)
	\fIDate, Time\fR		  (a NULL terminated string)
	[
	  \fIPlot title\fR 		  (a NULL terminated string)
	  \fINumber of variables\fR   (an int)
	  \fINumber of data points\fR (an int)
	  \fIflags\fR		  (a short)
	  \fIvariable header struct\fR (repeated numoutputs times)
		\fIvariable name\fR	  (a NULL terminated string)
		\fIvariable type\fR	  (an int)
	  \fIset of outputs\fR	  (repeated numpoints times)
	] repeated one or more times.
.fi
.PP
A set of outputs is a vector of doubles of length numoutputs, or
a vector of real-imaginary pairs of doubles if the data is complex.
.SH "SEE ALSO"
nutmeg(1), spice(1), writedata(3)
.SH AUTHOR
Wayne Christopher (faustus@cad.berkeley.edu)
.SH BUGS
If variable names and the title
and plotname strings have trailing
blanks in them they will be stripped off when the file is read, if
it is in ascii format.
.PP
If a plot title begins with "Title:" \fBnutmeg\fR will be fooled into thinking
that this is an ascii format file.
\fBSconvert\fR always requires the type to be specified, however.


NEWS
Ngspice-26, Jan 4th, 2014
============

- New features:
    + fft command may (optionally) use fftw3 library
    + add nint() rounding function
    + usage of 'temper' in behavioral devices
    + check for soa (safe operating area) in several device models
    + library processing rewritten completely to enhance ngspice
      compatibility (.lib)
    + include file processing rewritten to allow nesting and various
      absolute and relative path names (.inc)
    + asinh, acosh, atanh functions
    + shared ngspice option added: ngspice optionally compiles as
      a shared object or dynamic link library
    + use a hash table to massively speeding up finding a model
    + implement a very basic .if/.else block
    + implement a signed power function `pwr' for controlled sources
    + implement multiplier `m' in F, G source
    + apply Area Calculation Method (ACM) to the bsim3v3.3 model
    + implement `tc1', `tc2' instance parameters

- Bug fixes:
    + many, many ... please see
      http://sourceforge.net/p/ngspice/ngspice/ci/master/tree/
      and check the 'history' button on the upper right.
    + removed memory leaks


Ngspice-25, Jan 4th, 2013
============

- New features:
    + add temperature coefficients for B source
    + commands "mdump", "mrdump" to dump matrix to file
    + allow `time', `temper', and `hertz' in expressions with behavioral R, L, C
    + Enable ctrl-c to interrupt a simulation also for Windows GUI
    + tests/bsim4: use the standard CMC model qa check
    + implement exit code for command "quit"
    + .nodeset all = value
    + memristor code model in extradev
    + new scalable diode model including tunnel component
    + variable xtrtol may override the XSPICE reduction of trtol
    + XSPICE example: delta-sigma converter
    + XSPICE: new d_source model
    + new option noopac
    + bsim3: make additonal parameters accessible with @mxx[par]
    + enable Mac OS X compilation
    + option reseries = val
    + commands "snsave", "snload"
    + update hisim to version 2.7.0, new qa tests
    + hisimhv update to version 1.2.3, new qa tests
    + allow Ctrl-d to quit
    + jfet model temperature extension, tcv and bex parameter
    + speed up vector handling by hash table for vector address storage and retrieval
    + update PSS code


- Bug fixes:
    + many, many ... please see
      http://ngspice.git.sourceforge.net/git/gitweb.cgi?p=ngspice/ngspice;a=shortlog;pg=0
    + remove memory leaks
    + update tclspice for MS Windows
    + adms3 removed due to licensing issues


Ngspice-24, Jan 29th, 2012
============

- New features:
    + .csparam allows to tranfer a param into a vector in .control ... .endc
    + functions floor and ceil
    + syntax Gxxx n1 n2 TABLE {expression} = (x0, y0) (x1, y1) (x2, y2)
    + syntax Gxxx n1 n2 value={equation} or Exxx n1 n2 value={equation}
    + variable strict_errorhandling to exit ngspice with exit code 1
      if any error occurs during circuit parsing
    + command altermod allows reading from file
    + command "devhelp" to show available devices and describe device parameters
    + command "inventory" to print the number of istances of each device.
    + function cph(vec) delivers phase from a vector without jumps at +-PI
    + enable 64 bit compilation in MS Visual Studio
    + --adms3 flag to compile with (experimental) adms code
    + still very experimental pss code
    + command 'remcirc' to remove a circuit from the list
    + command 'wrdata' aknowledges appendwrite and prints complex variables
    + allow comment lines inside of command files
    + path variable NGSPICE_INPUT
    + 'Filesource' code model

- Bug fixes:
    + More on prototypes, type casts, reordering of functions, compiler warnings
    + fix a numerical problem in the hisim2 and hisimhv models
    + plot i(vdd) from a loaded rawfile
    + fix the guard for device generated internal nodes (via CKTmkVolt())
    + reduce memory leaks for command 'reset'
    + parameter processing for nested .subckt
    + bsim4/b4noi.c: correct init the correlated noise slot in noiseDens and lnNdens vector
    + Many small bugs

Ngspice-23, June 5th, 2011
============

- New features:
    + Reduction of trtol to 1 only if 'A' devices in the circuit and xspice is enabled
    + command wrs2p to write a s-parameter file using Touchstone vers. 1 format
    + d_source d_state d_ram, test cases for those xspice codemodels
    + transient noise simulation added to independent voltage and current sources
    + Random telegraph noise added to independent voltage and current sources
    + 'time = nnn' in stop command added
    + random voltage generator option trrandom to independent voltage and current sources
    + ngspice build possible in a separate directory (e.g. in ng-spice-rework/release)
    + update FIND .. WHEN measurements
    + diode & bjt temperature model update with tlev and tlevc
    + merge bsim3v1a code with bsim3v0 and bsim3v1s code with bsim3v1
    + bjt2: removed - all GP extensions of this model are now in the bjt model
    + HiSIM_HV model vers. 1.2.1 added
    + bsim3v32: delvto and mulu0 as instance parameters
    + Replace HiSIM1 by HiSIM_2.5.1_Release_20110407

- Bug fixes:
    + More on prototypes, type casts, reordering of functions, compiler warnings
    + allow models like 2N2904 or 2SK136 also in subcircuits
    + fix sensitvity calculation (for the dc-2+.cir testcase)
    + allow multiple call to setup w/ correct node collapsing
    + Many small bugs

Ngspice-22, Sept 26th, 2010
============

- New features:
    + OpenMP multicore support for BSIM3, BSIM4, and BSIMSOI4
    + reinstate {$var} expansion in interactive interpreter
    + .TITLE line added
    + update to 'spectrum' script
    + par('expression') in .four, .plot, .print, .meas, .save commands
    + command 'option' for use in spinit, .spiceinit, and in scripts
    + adms procedure updated
    + new random number generator, random functions sunif() and sgauss()

src/ngsconvert.c
/**********
Copyright 1990 Regents of the University of California.  All rights reserved.
Author: 1985 Wayne A. Christopher, U. C. Berkeley CAD Group
**********/

/*
 * Main routine for sconvert.
 */

#include "ngspice/ngspice.h"
#include <stdio.h>

#include "ngspice/fteinput.h"
#include "ngspice/cpdefs.h"
#include "ngspice/ftedefs.h"
#include "ngspice/sim.h"
#include "ngspice/suffix.h"
#include "frontend/display.h"
#include "../misc/mktemp.h"

FILE *cp_in = NULL;
FILE *cp_out = NULL;
FILE *cp_err = NULL;
FILE *cp_curin = NULL;
FILE *cp_curout = NULL;
FILE *cp_curerr = NULL;
int cp_maxhistlength;
bool cp_debug = FALSE;
char cp_chars[128];
bool cp_nocc = TRUE;
bool ft_stricterror = FALSE;
bool ft_parsedb = FALSE;
struct circ *ft_curckt = NULL;

char *cp_program = "sconvert";


#define tfread(ptr, siz, nit, fp)   if (fread((ptr), (siz), \
                        (nit), (fp)) != (nit)) { \
                fprintf(cp_err, "Error: unexpected EOF\n"); \
                    return (NULL); }

#define tfwrite(ptr, siz, nit, fp)  if (fwrite((ptr), (siz), \
                        (nit), (fp)) != (nit)) { \
                    fprintf(cp_err, "Write error\n"); \
                    return; }


void
Input(REQUEST *request, RESPONSE *response)
{
    switch (request->option) {
    case char_option:
	response->reply.ch = inchar(request->fp);
	response->option = request->option;
	break;
    default:
	/* just ignore, since we don't want a million error messages */
	response->option = error_option;
	break;
    }
    return;
}


static char *
fixdate(char *date)
{
    char buf[20];
    int i;

    (void) strcpy(buf, date);
    for (i = 17; i > 8; i--)
        buf[i] = buf[i - 1];
    buf[8] = ' ';
    buf[18] = '\0';
    return (copy(buf));
}


static struct plot *
oldread(char *name)
{
    struct plot *pl;
    char buf[BSIZE_SP];
    struct dvec *v, *end = NULL;
    short nv;       /* # vars */
    long np;        /* # points/var. */
    long i, j;
    short a;        /* The magic number. */
    float f1, f2;
    FILE *fp;

    if (!(fp = fopen(name, "r"))) {
        perror(name);
        return (NULL);
    }
    pl = alloc(struct plot);
    tfread(buf, 1, 80, fp);
    buf[80] = '\0';
    for (i = (int) strlen(buf) - 1; (i > 1) && (buf[i] == ' '); i--)
        ;
    buf[i + 1] = '\0';
    pl->pl_title = copy(buf);

    tfread(buf, 1, 16, fp);
    buf[16] = '\0';
    pl->pl_date = copy(fixdate(buf));

    tfread(&nv, sizeof (short), 1, fp);

    tfread(&a, sizeof (short), 1, fp);
    if (a != 4)
        fprintf(cp_err, "Warning: magic number 4 is wrong...\n");

    for (i = 0; i < nv; i++) {
        v = alloc(struct dvec);
        if (end)
            end->v_next = v;
        else
            pl->pl_scale = pl->pl_dvecs = v;
        end = v;
        tfread(buf, 1, 8, fp);
        buf[8] = '\0';
        v->v_name = copy(buf);
    }
    for (v = pl->pl_dvecs; v; v = v->v_next) {
        tfread(&a, sizeof (short), 1, fp);
        v->v_type = a;
    }

    /* If the first output variable is type FREQ then there is complex
     * data, otherwise the data is real.
     */
    i = pl->pl_dvecs->v_type;
    if ((i == SV_FREQUENCY) || (i == SV_POLE) || (i == SV_ZERO))
        for (v = pl->pl_dvecs; v; v = v->v_next)
            v->v_flags |= VF_COMPLEX;
    else 
        for (v = pl->pl_dvecs; v; v = v->v_next)
            v->v_flags |= VF_REAL;

    /* Check the node indices -- this shouldn't be a problem ever. */
    for (i = 0; i < nv; i++) {
        tfread(&a, sizeof(short), 1, fp);
        if (a != i + 1) 
	  fprintf(cp_err, "Warning: output %d should be %ld\n",
		  a, i);
    }
    tfread(buf, 1, 24, fp);
    buf[24] = '\0';
    pl->pl_name = copy(buf);
    /* Now to figure out how many points of data there are left in
     * the file. 
     */
    i = ftell(fp);
    (void) fseek(fp, 0L, SEEK_END);
    j = ftell(fp);
    (void) fseek(fp, i, SEEK_SET);
    i = j - i;
    if (i % 8) {    /* Data points are always 8 bytes... */
        fprintf(cp_err, "Error: alignment error in data\n");
        (void) fclose(fp);
        return (NULL);
    }
    i = i / 8;
    if (i % nv) {
        fprintf(cp_err, "Error: alignment error in data\n");
        (void) fclose(fp);
        return (NULL);
    }
    np = i / nv;

    for (v = pl->pl_dvecs; v; v = v->v_next) {
        v->v_length = (int) np;
        if (isreal(v)) {
            v->v_realdata = TMALLOC(double, np);
        } else {
            v->v_compdata = TMALLOC(ngcomplex_t, np);
        }
    }
    for (i = 0; i < np; i++) {
        /* Read in the output vector for point i. If the type is
         * complex it will be float and we want double.
         */
        for (v = pl->pl_dvecs; v; v = v->v_next) {
            if (v->v_flags & VF_REAL) {
                tfread(&v->v_realdata[i], sizeof (double),
                        1, fp);
            } else {
                tfread(&f1, sizeof (float), 1, fp);
                tfread(&f2, sizeof (float), 1, fp);
                realpart(v->v_compdata[i]) = f1;
                imagpart(v->v_compdata[i]) = f2;
            }
        }
    }
    (void) fclose(fp);
    return (pl);
}


static void
oldwrite(char *name, bool app, struct plot *pl)
{
    short four = 4, k;
    struct dvec *v;
    float f1, f2, zero = 0.0;
    char buf[80];
    int i, j, tp = VF_REAL, numpts = 0, numvecs = 0;
    FILE *fp;

    if (!(fp = fopen(name, app ? "a" : "w"))) {
        perror(name);
        return;
    }

    for (v = pl->pl_dvecs; v; v = v->v_next) {
        if (v->v_length > numpts)
            numpts = v->v_length;
        numvecs++;
        if (iscomplex(v))
            tp = VF_COMPLEX;
    }

    /* This may not be a good idea... */
    if (tp == VF_COMPLEX)
        pl->pl_scale->v_type = SV_FREQUENCY;

    for (i = 0; i < 80; i++)
        buf[i] = ' ';
    for (i = 0; i < 80; i++)
        if (pl->pl_title[i] == '\0')
            break;
        else
            buf[i] = pl->pl_title[i];
    tfwrite(buf, 1, 80, fp);

    for (i = 0; i < 80; i++)
        buf[i] = ' ';
    for (i = 0; i < 16; i++)
        if (pl->pl_date[i] == '\0')
            break;
        else
            buf[i] = pl->pl_date[i];
    tfwrite(buf, 1, 16, fp);

    tfwrite(&numvecs, sizeof (short), 1, fp);
    tfwrite(&four, sizeof (short), 1, fp);

    for (v = pl->pl_dvecs; v; v = v->v_next) {
        for (j = 0; j < 80; j++)
            buf[j] = ' ';
        for (j = 0; j < 8; j++)
            if (v->v_name[j] == '\0')
                break;
            else
                buf[j] = v->v_name[j];
        tfwrite(buf, 1, 8, fp);
    }

    for (v = pl->pl_dvecs; v; v = v->v_next) {
        j = (short) v->v_type;
        tfwrite(&j, sizeof (short), 1, fp);
    }

    for (k = 1; k < numvecs + 1; k++)
        tfwrite(&k, sizeof (short), 1, fp);
    for (j = 0; j < 80; j++)
        buf[j] = ' ';
    for (j = 0; j < 24; j++)
        if (pl->pl_name[j] == '\0')
            break;
        else
            buf[j] = pl->pl_name[j];
    tfwrite(buf, 1, 24, fp);
    for (i = 0; i < numpts; i++) {
        for (v = pl->pl_dvecs; v; v = v->v_next) {
            if ((tp == VF_REAL) && isreal(v)) {
                if (i < v->v_length) {
            tfwrite(&v->v_realdata[i], sizeof (double), 1, fp);
                } else {
    tfwrite(&v->v_realdata[v->v_length - 1], sizeof (double), 1, fp);
                }
            } else if ((tp == VF_REAL) && iscomplex(v)) {
                fprintf(cp_err, "internal error, everything real, yet complex ...\n");
                exit(1);
            } else if ((tp == VF_COMPLEX) && isreal(v)) {
                if (i < v->v_length)
                    f1 = (float) v->v_realdata[i];
                else
                    f1 = (float) v->v_realdata[v->v_length - 1];
                tfwrite(&f1, sizeof (float), 1, fp);
                tfwrite(&zero, sizeof (float), 1, fp);
            } else if ((tp == VF_COMPLEX) && iscomplex(v)) {
                if (i < v->v_length) {
                    f1 = (float) realpart(v->v_compdata[i]);
                    f2 = (float) imagpart(v->v_compdata[i]);
                } else {
                    f1 = (float) realpart(v->v_compdata[v-> v_length - 1]);
                    f2 = (float) imagpart(v->v_compdata[v-> v_length - 1]);
                }
                tfwrite(&f1, sizeof (float), 1, fp);
                tfwrite(&f2, sizeof (float), 1, fp);
            }
        }
    }

    (void) fclose(fp);
    return;
}


int
main(int ac, char **av)
{
    char *sf, *af;
    char buf[BSIZE_SP];
    char t, f;
    struct plot *pl;
    size_t n;
    char *infile = NULL;
    char *outfile = NULL;
    FILE *fp;
    cp_in = stdin;
    cp_out = stdout;
    cp_err = stderr;
    cp_curin = stdin;
    cp_curout = stdout;
    cp_curerr = stderr;

    switch (ac) {
        case 5: 
            sf = av[2];
            af = av[4];
            f = *av[1];
            t = *av[3];
            break;

        case 3:
            f = *av[1];
            t = *av[2];
            /* This is a pain, but there is no choice */
            sf = infile = smktemp("scin");
            af = outfile = smktemp("scout");
            if (!(fp = fopen(infile, "w"))) {
                perror(infile);
                exit(EXIT_BAD);
            }
            while ((n = fread(buf, 1, sizeof(buf), stdin)) != 0)
                (void) fwrite(buf, 1, n, fp);
            (void) fclose(fp);
            break;

        case 1: printf("Input file: ");
            (void) fflush(stdout);
            (void) fgets(buf, BSIZE_SP, stdin);
            sf = copy(buf);
            printf("Input type: ");
            (void) fflush(stdout);
            (void) fgets(buf, BSIZE_SP, stdin);
            f = buf[0];
            printf("Output file: ");
            (void) fflush(stdout);
            (void) fgets(buf, BSIZE_SP, stdin);
            af = copy(buf);
            printf("Output type: ");
            (void) fflush(stdout);
            (void) fgets(buf, BSIZE_SP, stdin);
            t = buf[0];
            break;
        default:
            fprintf(cp_err, 
                "Usage: %s fromtype fromfile totype tofile,\n",
                cp_program);
            fprintf(cp_err, "\twhere types are o, b, or a\n");
            fprintf(cp_err, 
                "\tor, %s fromtype totype, used as a filter.\n",
                cp_program);
            exit(EXIT_BAD);
    }
    switch(f) {
        case 'o' :
        pl = oldread(sf);
        break;

        case 'b' :
        case 'a' :
        pl = raw_read(sf);
        break;

        default:
        fprintf(cp_err, "Types are o, a, or b\n");
        exit(EXIT_BAD);
    }
    if (!pl)
        exit(EXIT_BAD);

    switch(t) {
        case 'o' :
        oldwrite(af, FALSE, pl);
        break;

        case 'b' :
        raw_write(af, pl, FALSE, TRUE);
        break;

        case 'a' :
        raw_write(af, pl, FALSE, FALSE);
        break;

        default:
        fprintf(cp_err, "Types are o, a, or b\n");
        exit(EXIT_BAD);
    }
    if (ac == 3) {
        /* Gotta finish this stuff up */
        if (!(fp = fopen(outfile, "r"))) {
            perror(outfile);
            exit(EXIT_BAD);
        }
        while ((n = fread(buf, 1, sizeof(buf), fp)) != 0)
            (void) fwrite(buf, 1, n, stdout);
        (void) fclose(fp);
        (void) unlink(infile);
        (void) unlink(outfile);
    }
    exit(EXIT_NORMAL);
}

void cp_pushcontrol(void) { }
void cp_popcontrol(void) { }
void out_init(void) { }
void cp_doquit(void) { exit(0); }
void cp_usrvars(struct variable **v1, struct variable **v2) { *v1 = NULL; *v2 = NULL;  return; }
int cp_evloop(char *s) { NG_IGNORE(s); return (0); }
void cp_ccon(bool o) { NG_IGNORE(o); }
char*if_errstring(int c) { NG_IGNORE(c); return ("error"); }
void out_printf(char *fmt, ...) { NG_IGNORE(fmt); }
void out_send(char *string) { NG_IGNORE(string); }
struct variable * cp_enqvar(char *word) { NG_IGNORE(word); return (NULL); }
struct dvec *vec_get(const char *word) { NG_IGNORE(word); return (NULL); }
void cp_ccom(wordlist *w, char *b, bool e) {
  NG_IGNORE(e);
  NG_IGNORE(b);
  NG_IGNORE(w); return; }
int cp_usrset(struct variable *v, bool i) {
    NG_IGNORE(i);
    NG_IGNORE(v); return(US_OK); }

int disptype;

